mirrord-layer/src/sockets.rs

```diff
@@ -0,0 +1,152 @@
+use std::{
+    collections::HashMap,
+    io::{Read, Write},
+    os::unix::io::AsRawFd,
+    sync::Mutex,
+};
+
+use multi_map::MultiMap;
+use queues::{IsQueue, Queue};
+use socketpair::{socketpair_stream, SocketpairStream};
+use tracing::{debug, error};
+
+pub struct ConnectionSocket {
+    pub read_fd: i32,
+    pub read_socket: SocketpairStream,
+    pub write_socket: SocketpairStream,
+    pub port: u16,
+}
+
+pub struct DataSocket {
+    pub connection_id: u16,
+    #[allow(dead_code)]
+    pub read_socket: SocketpairStream, /* Though this is unread, it's necessary to keep the
+                                        * socket open */
+    pub write_socket: SocketpairStream,
+}
+
+pub struct Sockets {
+    connections: Mutex<MultiMap<i32, u16, ConnectionSocket>>,
+    data: Mutex<HashMap<u16, DataSocket>>,
+    connection_queue: Mutex<Queue<u16>>,
+    pending_data: Mutex<HashMap<u16, Vec<u8>>>
+}
+
+impl Sockets {
+    pub fn new() -> Self {
```

Comment:
Minor thing, if `new` takes no arguments, then it should probably be an implementation of `default` instead.

```suggestion
impl Default for Sockets {
    fn default() -> Self { ... }
}
```